src/app-layer-imap.c

```diff
@@ -0,0 +1,606 @@
+            return true;
+        }
+
+        if ((command_start + 6) < input_len &&
+                __IMAPCompareBytesToCmd(input + command_start, "STATUS", 6)) {
+            return true;
+        }
+
+        if ((command_start + 7) < input_len &&
+                __IMAPCompareBytesToCmd(input + command_start, "ESEARCH", 7)) {
+            return true;
+        }
+
+        if ((command_start + 5) < input_len &&
+                __IMAPCompareBytesToCmd(input + command_start, "FLAGS", 5)) {
+            return true;
+        }
+    }
+    return false;
+}
+
+static bool IMAPServerResponseMailboxSize(const uint8_t *input, uint32_t input_len)
+{
+    /*mailbox size
+     * https://datatracker.ietf.org/doc/html/rfc9051#name-mailbox-size
+     * EXISTS
+     * These responses are always untagged.
+     * */
+
+    uint32_t mail_cnt_start = 0;
+    if ((mail_cnt_start = IMAPUntaggedCmd(input, input_len))) {
+        uint32_t command_start = 0;
+        if ((command_start = __IMAPMailCount(input, input_len, mail_cnt_start))) {
+            if ((command_start + 6) < input_len &&
+                    __IMAPCompareBytesToCmd(input + command_start, "EXISTS", 6)) {
+                return true;
+            }
+        }
+    }
+    return false;
+}
+
+static bool IMAPServerResponseMessageStatus(const uint8_t *input, uint32_t input_len)
+{
+    /*message status
+     * https://datatracker.ietf.org/doc/html/rfc9051#name-message-status
+     * EXPUNGE
+     * FETCH
+     * These responses are always untagged.
+     * */
+
+    uint32_t mail_cnt_start = 0;
+    if ((mail_cnt_start = IMAPUntaggedCmd(input, input_len))) {
+        uint32_t command_start = 0;
+        if ((command_start = __IMAPMailCount(input, input_len, mail_cnt_start))) {
+            if ((command_start + 7) < input_len &&
+                    __IMAPCompareBytesToCmd(input + command_start, "EXPUNGE", 7)) {
+                return true;
+            }
+
+            if ((command_start + 5) < input_len &&
+                    __IMAPCompareBytesToCmd(input + command_start, "FETCH", 5)) {
+                return true;
+            }
+        }
+    }
+    return false;
+}
+
+/**
+ * \brief Probe the input to see if it looks like imap check client commands.
+ *
+ * \retval ALPROTO_IMAP if it looks like imap, otherwise
+ *     ALPROTO_UNKNOWN.
+ */
+static AppProto IMAPProbingParserTs(
+        Flow *f, uint8_t direction, const uint8_t *input, uint32_t input_len, uint8_t *rdir)
+{
+    if (input_len < IMAP_MIN_FRAME_LEN) {
+        goto not_imap;
+    }
+
+    uint16_t line_end;
+    int ret = ByteExtractUint16(
+            &line_end, BYTE_BIG_ENDIAN, sizeof(uint16_t), (const uint8_t *)(input + input_len - 2));
+    if (ret == 0) {
+        goto not_imap;
+    }
+
+    if (line_end != 0x0d0a /*CRLF*/) {
+        goto not_imap;
+    }
+
+    /*
+     * DONE command without tag.
+     * */
+
+    if (input_len == 6 && __IMAPCompareBytesToCmd(input, "DONE", 4)) {
```

Comment:
why do we test for 6, then look for a 4 byte pattern?

Reply:
- for the "Done" command we should have line that is DONE followed by 2 bytes  0x0d0a CRLF.